mitiq/shadows/shadows_utils.py

```diff
@@ -0,0 +1,157 @@
+import mitiq
+import cirq
+import numpy as np
+from cirq.ops.pauli_string import PauliString
+from numpy.typing import NDArray
+import re
+
+
+def n_measurements_tomography_bound(epsilon: float, num_qubits: int) -> int:
+    """
+    This function returns the minimum number of classical shadows required
+    for state reconstruction for achieving the desired accuracy.
+
+    Args:
+        epsilon: The error on the estimator.
+        num_qubits: The number of qubits in the system.
+
+    Returns:
+        An integer that gives the number of snapshots required to satisfy the
+        shadow bound.
+    """
+    return int(34 * (4**num_qubits) * epsilon ** (-2))
+
+
+def local_clifford_shadow_norm(opt: PauliString[Any]) -> float:
+    """
+    Calculate shadow norm of an operator with random unitary sampled from local
+    Clifford group.
+    Args:
+        opt: a self-adjoint operator
+    Returns:
+        Shadow norm when unitary ensemble is local Clifford group.
+    """
+    norm = (
+        np.linalg.norm(
+            cirq.unitary(opt)
+            - np.trace(cirq.unitary(opt))
+            / 2 ** int(np.log2(cirq.unitary(opt).shape[0])),
+            ord=np.inf,
+        )
+        ** 2
+    )
+    return float(norm)
+
+
+def n_measurements_opts_expectation_bound(
+    error: float,
+    observables: List[PauliString[Any]],
+    failure_rate: float,
+) -> Tuple[int, int]:
+    """
+    This function returns the minimum number of classical shadows required and
+    the number of groups "k" into which we need to split the shadows for
+    achieving the desired accuracy and failure rate in operator expectation
+    value estimation.
+
+    Args:
+        error: The error on the estimator.
+        observables: List of cirq.PauliString corresponding to the
+            observables we intend to measure.
+        failure_rate: Rate of failure for the bound to hold.
+
+    Returns:
+        Integers quantifying the number of snapshots required to satisfy
+        the shadow bound and the chunk size required to attain the specified
+        failure rate.
+    """
+    M = len(observables)
+    K = 2 * np.log(2 * M / failure_rate)
+
+    N = (
+        34
+        * max(local_clifford_shadow_norm(o) for o in observables)
+        / error**2
+    )
+    return int(np.ceil(N * K)), int(K)
+
+
+def fidelity(
+    state_vector: NDArray[np.complex64],
+    rho: NDArray[np.complex64],
+) -> float:
+    """
+    Calculate the fidelity between a state vector and a density matrix.
+    Args:
+        state_vector: The vector whose norm we want to calculate.
+        rho: The operator whose norm we want to calculate.
+
+    Returns:
+        Scalar corresponding to the fidelity.
+    """
+    return np.reshape(state_vector.conj().T @ rho @ state_vector, -1).real[0]
+
+
+def transform_to_cirq_paulistring(
+    pauli_str: Union[str, mitiq.PauliString, cirq.PauliString[Any]]
+) -> Tuple[float, cirq.PauliString[Any]]:
```

Comment:
Why is this function needed? Is there a reason to convert to cirq `PauliString` objects instead of the mitiq `PauliString`?

Reply:
have deleted this function